### ISSUE TITLE
org.bouncycastle.asn1.DERApplicationSpecific cannot be cast to org.bouncycastle.asn1.ASN1Sequence

### ISSUE THREAD

**Issue by reejosamuel** [jruby/jruby-openssl#53]
I have a SOAP service with self signed SSL certificates, actually bumped on the error below when using savon. I opened an issue in that repository and they suggested I open an issue here.

The certificates where generated from a p12 file using the following commands

```
openssl pkcs12 -in qattbcap01.p12 -nocerts -out test01Key.pem
openssl pkcs12 -in qattbcap01.p12 -clcerts -nokeys -out test01Cert.pem
```

Using the following jRuby version:
`jruby 1.7.13 (1.9.3p392) 2014-06-24 43f133c on Java HotSpot(TM) 64-Bit Server VM 1.7.0_72-b14 [darwin-x86_64]`

```
require "httpi"

request = HTTPI::Request.new
request.url = "<removed url here>"

request.auth.ssl.cert_key_file     = Rails.root + "lib/test01Key.pem"
request.auth.ssl.cert_key_password = "test@1234"
request.auth.ssl.cert_file         = Rails.root + "lib/test01Cert.pem"
request.auth.ssl.verify_mode       = :none              
# request.auth.ssl.ssl_version       = :TLSv1             

HTTPI.get(request)
```

Getting the following error

```
Java::JavaLang::ClassCastException: org.bouncycastle.asn1.DERApplicationSpecific cannot be cast to org.bouncycastle.asn1.ASN1Sequence
    from org.jruby.ext.openssl.impl.PKey.readRSAPrivateKey(PKey.java:151)
    from org.jruby.ext.openssl.impl.PKey.readRSAPrivateKey(PKey.java:145)
    from org.jruby.ext.openssl.impl.PKey.readPrivateKey(PKey.java:101)
    from org.jruby.ext.openssl.PKey$PKeyModule.read(PKey.java:122)
    from org.jruby.ext.openssl.PKey$PKeyModule$INVOKER$s$0$1$read.call(PKey$PKeyModule$INVOKER$s$0$1$read.gen)
    from org.jruby.internal.runtime.methods.JavaMethod$JavaMethodN.call(JavaMethod.java:669)
    from org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:214)
    from org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:202)
    from org.jruby.ast.CallTwoArgNode.interpret(CallTwoArgNode.java:59)
    from org.jruby.ast.IfNode.interpret(IfNode.java:116)
    from org.jruby.ast.NewlineNode.interpret(NewlineNode.java:105)
    from org.jruby.ast.InstAsgnNode.interpret(InstAsgnNode.java:95)
    from org.jruby.ast.OpAsgnOrNode.interpret(OpAsgnOrNode.java:102)
    from org.jruby.ast.NewlineNode.interpret(NewlineNode.java:105)
    from org.jruby.evaluator.ASTInterpreter.INTERPRET_METHOD(ASTInterpreter.java:74)
    from org.jruby.internal.runtime.methods.InterpretedMethod.call(InterpretedMethod.java:139)
... 219 levels...
    from org.jruby.RubyKernel$INVOKER$s$1$0$require19.call(RubyKernel$INVOKER$s$1$0$require19.gen)
    from org.jruby.internal.runtime.methods.JavaMethod$JavaMethodOneOrNBlock.call(JavaMethod.java:350)
    from org.jruby.internal.runtime.methods.AliasMethod.call(AliasMethod.java:61)
    from org.jruby.internal.runtime.methods.AliasMethod.call(AliasMethod.java:61)
    from org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:326)
    from org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:170)
    from bin.rails.__file__(bin/rails:4)
    from bin.rails.load(bin/rails)
    from org.jruby.Ruby.runScript(Ruby.java:854)
    from org.jruby.Ruby.runScript(Ruby.java:847)
    from org.jruby.Ruby.runNormally(Ruby.java:716)
    from org.jruby.Ruby.runFromMain(Ruby.java:565)
    from org.jruby.Main.doRunFromMain(Main.java:395)
    from org.jruby.Main.internalRun(Main.java:290)
    from org.jruby.Main.run(Main.java:217)
    from org.jruby.Main.main(Main.java:197)jruby-1.7.13 :540 > 
```
##

**Comment by oozzal** [comment]
@reejosamuel, I get the same error. Any solutions @headius ?

```
Exception in thread "Ruby-0-Thread-9: /Users/uzzaldevkota/.rbenv/versions/jruby-9.0.5.0/lib/ruby/gems/shared/gems/sshkit-1.8.1/lib/sshkit/runners/parallel.rb:11" java.lang.ClassCas$
Exception: org.bouncycastle.asn1.DERApplicationSpecific cannot be cast to org.bouncycastle.asn1.ASN1Sequence
        at org.jruby.ext.openssl.impl.PKey.readRSAPrivateKey(PKey.java:151)
        at org.jruby.ext.openssl.impl.PKey.readRSAPrivateKey(PKey.java:145)
        at org.jruby.ext.openssl.impl.PKey.readPrivateKey(PKey.java:101)
        at org.jruby.ext.openssl.PKey$PKeyModule.read(PKey.java:119)
        at org.jruby.ext.openssl.PKey$PKeyModule$INVOKER$s$0$1$read.call(PKey$PKeyModule$INVOKER$s$0$1$read.gen)
        at org.jruby.internal.runtime.methods.JavaMethod$JavaMethodN.call(JavaMethod.java:729)
        at org.jruby.runtime.callsite.CachingCallSite.callBlock(CachingCallSite.java:77)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:83)
        at org.jruby.ir.instructions.CallBase.interpret(CallBase.java:414)
        at org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:348)
        at org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:77)
        at org.jruby.internal.runtime.methods.MixedModeIRMethod.INTERPRET_METHOD(MixedModeIRMethod.java:140)
        at org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:126)
        at org.jruby.runtime.callsite.CachingCallSite.callBlock(CachingCallSite.java:77)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:83)
        at org.jruby.ir.instructions.CallBase.interpret(CallBase.java:414)
        at org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:348)
        at org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:77)
        at org.jruby.internal.runtime.methods.MixedModeIRMethod.INTERPRET_METHOD(MixedModeIRMethod.java:140)
        at org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:126)
        at org.jruby.runtime.callsite.CachingCallSite.callBlock(CachingCallSite.java:77)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:83)
        at org.jruby.ir.instructions.CallBase.interpret(CallBase.java:414)
        at org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:348)
        at org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:77)
        at org.jruby.ir.interpreter.Interpreter.INTERPRET_BLOCK(Interpreter.java:132)
        at org.jruby.runtime.MixedModeIRBlockBody.commonYieldPath(MixedModeIRBlockBody.java:148)
        at org.jruby.runtime.IRBlockBody.doYield(IRBlockBody.java:179)
        at org.jruby.runtime.BlockBody.yield(BlockBody.java:96)
        at org.jruby.runtime.Block.yield(Block.java:166)
        at org.jruby.RubyArray.collect(RubyArray.java:2286)
        at org.jruby.RubyArray.map19(RubyArray.java:2300)
...
```

**Comment by reejosamuel** [comment]
@oozzal Just revisit how key and certs are generated from p12 etc. 
My issue was resolved once I re-did everything.

**Comment by kares** [comment]
defunct (expected to be resolved by now >= 0.10.6), closing
